### ISSUE TITLE
Trivial tests fail when using Microsoft SQL Server

### ISSUE THREAD

**Issue by raichoo** [tpolecat/doobie#218]
I'm trying to use the testing capabilities of Doobie to check Microsoft SQL queries but
even simple queries like the one below do now pass the tests.

    object MyTestSpec extends Specification with AnalysisSpec {
      val transactor = DriverManagerTransactor[Task](
          "com.microsoft.sqlserver.jdbc.SQLServerDriver",
          "jdbc:sqlserver://…;databaseName=…",
          "…",
          "…"
        )

      val q = sql"""SELECT 42""".query[Int]

      check(q)
    }

Output:

    [info] Query0[Int] defined at DBTests.scala:15
    [info]   SELECT 42
    [error]   x SQL Compiles and Typechecks
    [error]    x doobie.enum.jdbctype$JdbcType: invalid ordinal: -155 (specs2.scala:61)
    [info]
    [info]
    [info] Total for specification MyTestSpec
    [info] Finished in 18 ms
    [info] 1 example, 1 failure, 0 error
    [info]
    [info] ScalaTest
    [info] Run completed in 3 seconds, 662 milliseconds.
    [info] Total number of tests run: 0
    [info] Suites: completed 0, aborted 0
    [info] Tests: succeeded 0, failed 0, canceled 0, ignored 0, pending 0
    [info] No tests were executed.

Kind regards,
raichoo

**Comment by tpolecat** [comment]
MS-SQL is reporting that the column has type `-155`, which outside the JDBC specification. Can you do this and report the results? I already have a special case for `NVarChar` being reported as `-10` so I can probably just expand the case statement a bit more.

```scala
scala> import doobie.imports._
scala> import doobie.free.{ databasemetadata => DMD }
scala> import scalaz._, Scalaz._, scalaz.concurrent.Task

scala> val xa: Transactor[Task] = ...

scala> val m = HC.getMetaData(DMD.getTypeInfo.flatMap(DMD.liftResultSet(_, HRS.list[(String,Int)]))).transact(xa).run
m: List[(String, Int)] = ...

scala> m.sortBy(_._2).foreach(Console.println)
(bool,-7)
(bit,-7)
(int8,-5)
(bigserial,-5)
(oid,-5)
(bytea,-2)
(char,1)
(bpchar,1)
...
```

**Comment by raichoo** [comment]
Here is the output:

    (datetimeoffset,-155)
    (sql_variant,-150)
    (xml,-16)
    (ntext,-16)
    (nchar,-15)
    (nvarchar,-9)
    (sysname,-9)
    (bit,-7)
    (tinyint,-6)
    (tinyint identity,-6)
    (bigint,-5)
    (bigint identity,-5)
    (image,-4)
    (varbinary,-3)
    (binary,-2)
    (timestamp,-2)
    (text,-1)
    (uniqueidentifier,1)
    (char,1)
    (numeric,2)
    (numeric() identity,2)
    (decimal,3)
    (money,3)
    (smallmoney,3)
    (decimal() identity,3)
    (int,4)
    (int identity,4)
    (smallint,5)
    (smallint identity,5)
    (real,7)
    (float,8)
    (varchar,12)
    (date,91)
    (time,92)
    (datetime2,93)
    (datetime,93)
    (smalldatetime,93)
    (datetimeoffset,-155)
    (sql_variant,-150)
    (xml,-16)
    (ntext,-16)
    (nchar,-15)
    (nvarchar,-9)
    (sysname,-9)
    (bit,-7)
    (tinyint,-6)
    (tinyint identity,-6)
    (bigint,-5)
    (bigint identity,-5)
    (image,-4)
    (varbinary,-3)
    (binary,-2)
    (timestamp,-2)
    (text,-1)
    (uniqueidentifier,1)
    (char,1)
    (numeric,2)
    (numeric() identity,2)
    (decimal,3)
    (money,3)
    (smallmoney,3)
    (decimal() identity,3)
    (int,4)
    (int identity,4)
    (smallint,5)
    (smallint identity,5)
    (real,7)
    (float,8)
    (varchar,12)
    (date,91)
    (time,92)
    (datetime2,93)
    (datetime,93)
    (smalldatetime,93)

**Comment by tpolecat** [comment]
Ok, so `datetimeoffset/-155` (the inferred type of literal `42` for some reason) and `sql_variant/-150` have ordinals that are outside the spec. I'll see if I can find some doc on what these types really mean and with some luck I can just map them to proper JDBC constants (doobie already maps H2's `nvarchar/-10` to the spec's `-9` with no ill effect, heh-heh, so hopefully something similar will work here).

Thanks! I'll let you know when a snapshot is available to play with.

**Comment by tpolecat** [comment]
Ok per the [doc](https://msdn.microsoft.com/en-us/library/ms378878(v=sql.110).aspx):

> The SQL Server `sqlvariant` data type is not currently supported by the JDBC driver. If a query is used to retrieve data from a table that contains a column of the `sqlvariant` data type, an exception will occur.

So I'm not going to worry about that one.

As for `datetimeoffset` we have the constant `microsoft.sql.Types.DATETIMEOFFSET`, mapping to type `microsoft.sql.DateTimeOffset`, which is a date+time with UTC offset. Odd that this is the inferred type for an integer literal.

@raichoo do you have better luck if you do a select from an actual table with declared column types?

**Comment by tpolecat** [comment]
Aaaah nevermind, I know what's going on. It's not inferring `DATETIMEOFFSET` for `42`, it's just having trouble loading up the schema metadata. Can fix. Sorry for the confusion.

**Comment by tpolecat** [comment]
Reopening pending confirmation. @raichoo can you try `0.2.3-SNAPSHOT` (on sonatype snapshots) and see if it resolves your problem?

**Comment by raichoo** [comment]
Hey,

I tested some queries with the snapshot and it seems to work so far. However the trivial one still fails as well as queries containing CTEs here are some failing examples:

    val trivial = sql"""SELECT 42""".query[Int]

    val cte =
          sql"""WITH cte_test AS (
                SELECT 42::int AS "FOO", 43::int AS "BAR"
              ) SELECT * FROM cte_test WHERE "FOO" = 42
           """.query[(Int, Int)]

    check(trivial)
    check(cte)

Fails like so:

    [info] Query0[Int] defined at DBTests.scala:18
    [info]   SELECT 42
    [error]   x SQL Compiles and Typechecks
    [error]    x com.microsoft.sqlserver.jdbc.SQLServerException: Unable to identify the table
    [error]      SELECT 42 for the metadata. (specs2.scala:61)
    [info]
    [info] Query0[(Int, Int)] defined at DBTests.scala:21
    [info]   WITH cte_test AS (
    [info]   SELECT 42::int AS "FOO", 43::int AS "BAR"
    [info]   ) SELECT * FROM cte_test WHERE "FOO" = 42
    [info]
    [error]   x SQL Compiles and Typechecks
    [error]    x com.microsoft.sqlserver.jdbc.SQLServerException: Unable to identify the table
    [error]      WITH cte_test AS (
    [error]     SELECT 42::int AS "FOO", 43::int AS "BAR"
    [error]     ) SELECT * FROM
    [error]      cte_test WHERE "FOO" = 42
    [error]     for the metadata. (specs2.scala:61)


Kind regards,
raichoo

**Comment by tpolecat** [comment]
Hm, that failure means the DB couldn't prepare the statement. Is that valid SQL?

**Comment by tpolecat** [comment]
I'm closing this issue since the type mapping issue is fixed, but see #222 for a continuation re: the metadata problem.
